transport_nantes/photo/views.py

```diff
@@ -85,3 +91,257 @@ def get_context_data(self, **kwargs):
+
+        # If the user is logged in but never voted, we use the form with
+        # consent
+        elif not Vote.objects.filter(user=user).exists() and user:
+            self.form_class = SimpleVoteFormWithConsent
+
+        return super().post(request, *args, **kwargs)
+
+    def form_valid(self, form):
+        """Handle the forms submissions
+
+        This view can have 3 different forms, all targeted to create a vote
+        event and optionally subscribe user to a mailing list.
+
+        The 3 forms are:
+        - AnonymousVoteForm: Used when the user is not logged in and has not
+        voted yet. The form is used to create a vote and subscribe the user
+        to the mailing list if they agree to it.
+        - SimpleVoteForm: Used when the user is logged in or has already
+        voted with this session. User was already asked for consent at this
+        point and it's not asked again.
+        - SimpleVoteFormWithConsent: Used when the user is logged in but
+        never voted. The form is used to create a vote and subscribe the user
+        to the mailing list if they agree to it. The difference is that this
+        form doesn't ask for email or captcha compared to AnonymousVoteForm.
+
+        """
+        # The photo sha1 is included in the URL
+        photo_sha1 = self.kwargs.get("photo_sha1")
+        photo = get_object_or_404(PhotoEntry, sha1_name=photo_sha1)
+        if not photo.accepted:
+            return HttpResponseForbidden("Photo not accepted")
+
+        if not self.request.user.is_authenticated:
+            # email address is only asked in anonymous vote form
+            email_address = self.request.POST.get("email_address", None)
+            if email_address:
+                user = User.objects.get_or_create(email=email_address)[0]
+            else:
+                # When the user is anon and has already voted previously,
+                # we don't ask for email address
+                user = None
+        else:
+            user = self.request.user
+
+        # Consent is asked in both anonymous and logged in vote forms
+        # only while the user hasn't voted yet. Consent box is opt-in :
+        # if the user doesn't check it, we subscribe them to the mailing
+        # list.
+        no_consent = self.request.POST.get("consent_box", None)
+        if not no_consent and user:
+            try:
+                mailing_list = MailingList.objects.get(
+                    mailing_list_token="operation-pieton")
+                subscribe_user_to_list(user, mailing_list)
+                logger.info(
+                    f"Subscribed user {user.email} to list {mailing_list}.")
+            except MailingList.DoesNotExist:
+                logger.error(
+                    "Mailing list operation-pieton does not exist")
+
+        # Vote nature is a constant among the forms
+        vote_value = self.request.POST.get("vote_value", None)
+        if vote_value == "upvote":
```

Comment:
Why doesn't the form set the correct True/False for you from the outset?  Not a huge deal, just wondering...

More pertinent, the pythonic way do this would be to have a dict that does the mapping, you do an assignment through the dict and on key error you return the appropriate response.  That response is _not_ 403, which means you understood the response but refuse to authorise it.

Reply:
Javascript sends the string "true" for `True` in it POST data, it was easier and more idiomatic for me to write out "upvote".